common/client/node_lifecycle.go

```diff
@@ -0,0 +1,431 @@
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePolls = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_total",
+		Help: "The total number of poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePollsFailed = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_failed",
+		Help: "The total number of failed poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePollsSuccess = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_success",
+		Help: "The total number of successful poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+)
+
+// zombieNodeCheckInterval controls how often to re-check to see if we need to
+// state change in case we have to force a state transition due to no available
+// nodes.
+// NOTE: This only applies to out-of-sync nodes if they are the last available node
+func zombieNodeCheckInterval(noNewHeadsThreshold time.Duration) time.Duration {
+	interval := noNewHeadsThreshold
+	if interval <= 0 || interval > QueryTimeout {
+		interval = QueryTimeout
+	}
+	return utils.WithJitter(interval)
+}
+
+func (n *node[CHAIN_ID, HEAD, RPC]) setLatestReceived(blockNumber int64, totalDifficulty *utils.Big) {
+	n.stateMu.Lock()
+	defer n.stateMu.Unlock()
+	n.stateLatestBlockNumber = blockNumber
+	n.stateLatestTotalDifficulty = totalDifficulty
+}
+
+const (
+	msgCannotDisable = "but cannot disable this connection because there are no other RPC endpoints, or all other RPC endpoints are dead."
+	msgDegradedState = "Chainlink is now operating in a degraded state and urgent action is required to resolve the issue"
+)
+
+// Node is a FSM
+// Each state has a loop that goes with it, which monitors the node and moves it into another state as necessary.
+// Only one loop must run at a time.
+// Each loop passes control onto the next loop as it exits, except when the node is Closed which terminates the loop permanently.
+
+// This handles node lifecycle for the ALIVE state
+// Should only be run ONCE per node, after a successful Dial
+func (n *node[CHAIN_ID, HEAD, RPC]) aliveLoop() {
+	defer n.wg.Done()
+
+	{
+		// sanity check
+		state := n.State()
+		switch state {
+		case nodeStateAlive:
+		case nodeStateClosed:
+			return
+		default:
+			panic(fmt.Sprintf("aliveLoop can only run for node in Alive state, got: %s", state))
+		}
+	}
+
+	noNewHeadsTimeoutThreshold := n.noNewHeadsThreshold
+	pollFailureThreshold := n.nodePoolCfg.PollFailureThreshold()
+	pollInterval := n.nodePoolCfg.PollInterval()
+
+	lggr := n.lfcLog.Named("Alive").With("noNewHeadsTimeoutThreshold", noNewHeadsTimeoutThreshold, "pollInterval", pollInterval, "pollFailureThreshold", pollFailureThreshold)
+	lggr.Tracew("Alive loop starting", "nodeState", n.State())
+
+	headsC := make(chan HEAD)
+	sub, err := n.rpc.Subscribe(n.nodeCtx, headsC, "newHeads")
```

Comment:
Feels like "newHeads" argument is too dependent on evm RPC implementation. Can we replace 
`Subscribe(ctx context.Context, channel chan<- HEAD, args ...interface{}) (types.Subscription, error)` with 
`SubscribeNewHeads(ctx context.Context, channel chan<- HEAD) (types.Subscription, error)`

P.S it would be cool to return `chan HEAD` instead of accepting it as argument, and close it to signal critical errors

Reply:
During the investigation phase, we hardly found any non-EVM chains supporting block subscriptions. We even discussed the possibility of implementing it ourselves with a polling mechanism under the hood. With that in mind, I think it's ok for now to go with a design that it's more EVM-like and change it once we find the need to do so.